### PR TITLE
Introduce Agent API middleware

### PR DESCRIPTION
**Pull request by azdagron**
This change introduces a middleware layer into the agent, which reduces some of the boilerplate present in the API handlers.

Specifically, the middleware:
- Handles security header enforcement
- Embellishes loggers with caller information (RPC being called, PID of caller, etc).
- Records telemetry for the RPC call and maintains global per-service connection counts
- Consolidates peertracker interaction to a single layer

Beyond the addition of the middleware, it provides:
- Exhaustive unit-testing of the endpoints package
- Cleans up some unused configuration for the endpoints package
- Moves the middleware infrastructure to a common package and provides aliases where appropriate (so the server implementation doesn't have to import multiple packages with the same name)

### CODE REVIEW

**Review by amartinezfayo — COMMENTED** [review]
Thanks @azdagron for doing this! It's really a step forward to simplify and improve the code.
All changes look good to me. I only have a couple of very minor comments.

File: pkg/server/api/rpccontext/alias.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+func WithCallCounter(ctx context.Context, counter api.CallCounter) context.Context {
+	return rpccontext.WithCallCounter(ctx, counter)
+}
+
+func CallCounter(ctx context.Context) api.CallCounter {
+	return rpccontext.CallCounter(ctx)
+}
+
+func AddMetricsLabel(ctx context.Context, name, value string) {
+	CallCounter(ctx).AddLabel(name, value)
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/server/api/rpccontext/alias.go:28]: The `AddLabel` method name was in the context of a `CallCounter` where it being a label for a "metric" is implied.

Although, turns out, these methods aren't used. I think we anticipated wanting to add labels to the call counters but it hasn't turned out to be the case. How would you feel if I opened a separate PR afterwards to remove them?

[the remark this answers is not in the document]
Comment (amartinezfayo) [reply pkg/server/api/rpccontext/alias.go:28]: If we are not using them, I would vote for just remove them. Doing that in a separate PR looks good to me.

**Review by amartinezfayo — APPROVED** [review]
🚢

**Pull request merged**